Add whitelist to limit openstack loadbalancers

hmorikaw opened the pull request:
I tried to deploy OpenStack Discoverer, but sync cycle was too long on a OpenStack cluster which has many projects.

My suggestion is to implement whitelist for rapid sync cycle.
```
$ kubectl logs openstack-discoverer-5556b6dcbc-ptj6n -n gimbal-discovery | grep 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'

time="2018-10-28T22:20:33Z" level=info msg="Successfully handled: update service 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'"
time="2018-10-28T22:51:58Z" level=info msg="Successfully handled: update endpoints 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'"
time="2018-10-28T22:51:59Z" level=info msg="Successfully handled: update service 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'"
time="2018-10-28T23:23:19Z" level=info msg="Successfully handled: update endpoints 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'"
time="2018-10-28T23:54:43Z" level=info msg="Successfully handled: update service 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'"
time="2018-10-28T23:54:43Z" level=info msg="Successfully handled: update endpoints 'dreco-001/openstack-626c5bf1-3cf8-438a-b6a1-e20419126eee'"
```

issue: #241 

Signed-off-by: Hiroaki Morikawa <hmorikaw@yahoo-corp.jp>

alexbrand commented on discovery/pkg/openstack/reconciler.go:
```diff
@@ -116,6 +118,24 @@ func (r *Reconciler) reconcile() {
+
+	// import white list
+	tmp := projects[:0]
+	if whitelist, ok := os.LookupEnv("PROJECT_WHITELIST"); ok && len(whitelist) > 0 {
```
I think this check should be performed in main.go, and then the list of projects flowed/injected into a field on the `Reconciler` type. 

That will make it more testable.

hmorikaw replied:
Thank you for your reply.
I change this imports env variable "WATCHED_PROJECTS", which is renamed, from main.go.

alexbrand commented:
Now that it is main, I am torn between using a flag or an env variable for the list of projects.

The environment variables that we currently have are in place because that is how most openstack tools expect their configuration.

Given that the list of projects is more of a gimbal concern, I am leaning towards using a flag to pass the list of projects, instead of an env variable.

I also think I prefer the idea of the whitelist. In the current implementation, we still query the openstack API for a list of all projects. Thus, the idea of a whitelist makes sense. That is, we will ask for all projects, but only reconcile the ones that appear in the whitelist. 

Maybe something like this in main.go:
```
 	flag.StringVar(&openstackProjectWhitelist, "openstack-project-whitelist", "", "Whitelist of projects that should be reconciled. If empty, load balancers across all projects will be reconciled.")
```

What do you all think?

alexbrand commented on discovery/pkg/openstack/reconciler.go:
```diff
@@ -116,6 +119,25 @@ func (r *Reconciler) reconcile() {
+				}
+			}
+		}
+		projects = tmp
```
This will cause problems if the user provides a project that doesn't exist.

alexbrand replied:
I think I would have taken a slightly different approach here. Instead of overwriting the projects we got from the OpenStack API, I would have added a check in the for loop (line 138):

```
for _, project := range projects {
  // skip the project if it is not in the whitelist
  if !contains(whitelist, project) { // pseudo-code here... this checks whether the project is in the whitelist
    continue
  }
  ...
  ...
```

hmorikaw replied:
@alexbrand I applied your approach. Please check my commit.

alexbrand commented:
@hmorikaw Please make sure to signoff your commits. More info here: https://github.com/heptio/gimbal/blob/master/CONTRIBUTING.md#dco-sign-off

hmorikaw commented:
I signed off my commits and changed the source of whitelist from environment variable to flag option.

stevesloka commented:
@hmorikaw I think this is looking better. Could you add some tests to cover the new functionality? Also, would be great to update docs to reference the new params.

hmorikaw commented:
@stevesloka I added a new test code 'discovery/pkg/openstack/reconciler_test.go' 
and updated the docs 'docs/openstack-discoverer.md' to reflect the new param.

stevesloka commented on docs/openstack-discoverer.md:
```diff
@@ -57,6 +57,7 @@ Arguments are available to customize the discoverer, most have defaults but othe
 | prometheus-listen-address | 8080 | The address to listen on for Prometheus HTTP requests
 | gimbal-client-qps | 5 | The maximum queries per second (QPS) that can be performed on the Gimbal Kubernetes API server
 | gimbal-client-burst | 10 | The maximum number of queries that can be performed on the Gimbal Kubernetes API server during a burst
+| openstack-project-whitelist | "" | The whitelist of projects that should be reconciled. If empty, load balancers across all projects will be reconciled.
```
nit: Could we add a small note that this list should be comma seperated since that's what the code is looking for

stevesloka commented:
Hey @hmorikaw, there's still one change to update with the name of the new param:

> My only thing would be to rename whitelist to openstack-project-watchlist: List of projects to be watched for reconcilation. If empty, load balancers across all projects will be reconciled

hmorikaw commented:
@stevesloka I'm so sorry. I rename `whitelist` to `watchlist`.

stevesloka commented:
/lgtm 

Thanks for the PR @hmorikaw, this is a really nice feature!
